packages/react-shadow/src/renderers/createShadowDOMRenderer.ts

```diff
@@ -0,0 +1,204 @@
+import type { GriffelRenderer, StyleBucketName } from '@griffel/core';
+
+import { createFallbackRenderer } from './createFallbackRenderer';
+
+const SUPPORTS_CONSTRUCTABLE_STYLESHEETS: boolean = (() => {
+  try {
+    new CSSStyleSheet();
+    return true;
+  } catch {
+    return false;
+  }
+})();
+
+type ExtendedCSSStyleSheet = CSSStyleSheet & {
+  bucketName: StyleBucketName;
+  metadata: Record<string, unknown>;
+};
+type GriffelShadowDOMRenderer = GriffelRenderer & {
+  adoptedStyleSheets: ExtendedCSSStyleSheet[];
+};
+
+let rendererId = 0;
+
+const styleBucketOrderingMap = styleBucketOrdering.reduce((acc, cur, j) => {
+  acc[cur as StyleBucketName] = j;
+  return acc;
+}, {} as Record<StyleBucketName, number>);
+
+function findInsertionPoint(
+  renderer: GriffelShadowDOMRenderer,
+  styleSheet: ExtendedCSSStyleSheet
+): ExtendedCSSStyleSheet | null {
+  let styleSheets = renderer.adoptedStyleSheets;
+  const targetOrder = styleBucketOrderingMap[styleSheet.bucketName];
+
+  let comparer = (sheet: ExtendedCSSStyleSheet): number =>
+    targetOrder - styleBucketOrderingMap[sheet.bucketName];
+
+  if (styleSheet.bucketName === 'm' && styleSheet.metadata) {
+    const mediaElements = renderer.adoptedStyleSheets.filter(
+      (styleSheet) => styleSheet.bucketName === 'm'
+    );
+
+    if (mediaElements.length) {
+      styleSheets = mediaElements;
+      comparer = (sheet) =>
+        renderer.compareMediaQueries(
+          styleSheet.metadata['m'] as string,
+          sheet.metadata['m'] as string
+        );
+    }
+  }
+
+  const length = styleSheets.length;
+  let index = length - 1;
+
+  while (index >= 0) {
+    const styleElement = styleSheets[index];
+
+    if (comparer(styleElement) > 0) {
+      return styleElement;
+    }
+
+    index--;
+  }
+
+  if (length > 0) {
+    return styleSheets[0];
+  }
+
+  return null;
+}
+
+function getCSSStyleSheetForBucket(
+  cssSheetsCache: Record<string, ExtendedCSSStyleSheet>,
+
+  bucketName: StyleBucketName,
+  metadata: Record<string, unknown> = {},
+
+  onStyleSheetInsert: (stylesheet: ExtendedCSSStyleSheet) => void
+): ExtendedCSSStyleSheet {
+  const isMediaBucket = bucketName === 'm';
+  const styleSheetKey: StyleBucketName | string = isMediaBucket
+    ? ((bucketName + metadata['m']) as string)
+    : bucketName;
+
+  if (!cssSheetsCache[styleSheetKey]) {
+    const styleSheet = new CSSStyleSheet() as ExtendedCSSStyleSheet;
+
+    styleSheet.bucketName = bucketName;
+    styleSheet.metadata = metadata;
+
+    cssSheetsCache[styleSheetKey] = styleSheet;
+    onStyleSheetInsert(styleSheet);
+  }
+
+  return cssSheetsCache[styleSheetKey];
+}
+
+function insertBefore<T extends CSSStyleSheet | ExtendedCSSStyleSheet>(
+  arr: T[],
+  sheetToInsert: T,
+  targetSheet: T | null
+): T[] {
+  if (targetSheet === null) {
+    return [...arr, sheetToInsert];
+  }
+
+  const index = arr.indexOf(targetSheet);
+
+  return [...arr.slice(0, index + 1), sheetToInsert, ...arr.slice(index + 1)];
```

Comment:
```suggestion
  return arr.splice(index, 0, sheetToInsert);
```

should do the same ting

Reply:
Since the return value is set on `.adoptedStyleSheets` this function must return a new array reference.

https://developer.mozilla.org/en-US/docs/Web/API/Document/adoptedStyleSheets#append_a_new_stylesheet

Reply:
This is a requirement for Chrome below 99, after that version `.adoptedStyleSheets` are mutable.